.github/workflows/ci.yml

```diff
@@ -0,0 +1,89 @@
+name: CI
+
+on:
+  push:
+    branches: [ main ]
+  pull_request:
+    branches: [ main ]
+  workflow_dispatch:
+
+permissions:
+  contents: read
+
+concurrency:
+  group: ci-${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  lint:
+    name: Lint (ruff + mypy)
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Set up Python 3.12
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.12"
+          cache: "pip"
+
+      - name: Install dependencies for linting
+        run: |
+          python -m pip install -U pip setuptools wheel
+          python -m pip install mypy ruff pandas-stubs "numpy>=1.26.3,<=2.1.3" matplotlib xgboost pandas scikit-learn openpyxl jinja2 python-ternary pytest shap interpret types-tqdm types-Pillow scikit-image lightgbm lime opencv-python-headless==4.10.0.84
+
+      - name: Run mypy
+        run: mypy .
+
+      - name: Run ruff
+        run: ruff check .
+
+  functional_test:
+    name: Functional test (python main.py)
+    runs-on: ubuntu-latest
+    needs: [lint]
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Set up Python 3.12
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.12"
+          cache: "pip"
+
+      - name: Install runtime dependencies
+        run: |
+          python -m pip install -U pip setuptools wheel
+          python -m pip install mypy ruff pandas-stubs "numpy>=1.26.3,<=2.1.3" matplotlib xgboost pandas scikit-learn openpyxl jinja2 python-ternary pytest shap interpret types-tqdm types-Pillow scikit-image lightgbm lime opencv-python-headless==4.10.0.84
+
+      - name: Run end-to-end script
+        run: python main.py
+
+  unit_tests:
+    name: Unit tests (pytest on neat_ml)
+    runs-on: ubuntu-latest
+    needs: [lint]
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Set up Python 3.12
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.12"
+          cache: "pip"
+
+      - name: Install test dependencies and package
+        run: |
+          python -m pip install -U pip setuptools wheel
+          python -m pip install mypy ruff pandas-stubs "numpy>=1.26.3,<=2.1.3" matplotlib xgboost pandas scikit-learn openpyxl jinja2 python-ternary pytest shap interpret types-tqdm types-Pillow scikit-image lightgbm lime opencv-python-headless==4.10.0.84
+          python -m pip install -v .
+
+      - name: Run pytest (import package)
+        run: |
+          cd /tmp
+          python -m pytest --pyargs neat_ml
```

Comment:
It looks like @adamwitmer didn't mention that a `ci.yml` file was already drafted in gh-6, since it was listed as a checklist item over there... so maybe he didn't actually read through this PR then.

Anyway, the merged CI configuration I added in gh-9 excludes the problematic functional test that is still present here, tests on more versions of Python and also tests on MacOS, and it uses a matrix setup to avoid duplication in the configuration, so I'll probably rebase this branch and keep the new CI config I added from there instead.